infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java

```diff
@@ -49,24 +45,48 @@ public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
      * @param r   The router that took in the message.
      */
     public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        if (shutdown) {
+        if (getState() == ServerState.SHUTDOWN) {
             log.warn("Server received {} but is already shutdown.", msg.getMsgType().toString());
             return;
         }
 
+        if (!isServerReadyToHandleMsg(msg)) {
+            r.sendResponse(ctx, msg, CorfuMsgType.NOT_READY.msg());
+            return;
+        }
+
         if (!getHandler().handle(msg, ctx, r)) {
             log.warn("Received unhandled message type {}", msg.getMsgType());
         }
     }
 
+    protected void setState(ServerState state) {
```

Comment:
Isn't there a race possible here?
You need to compareAndSet to check it's not shutdown.

Reply:
fixed